add picard iteration gmg

tcclevenger opened the pull request:
One can now use GMG with Picard iteration. There is an `Assert(newton_derivative_scaling_factor==0)` in `solve_iterated_advection_and_newton_stokes()` in `solver_schemes.cc` so that the computation is aborted if the code switches to Newton solver (still not implemented).

I've created a test based on `test/nonlinear_channel_flow_velocities_Newton_Stokes`, except the `.prm` has been changed to use harmonic averaging and to never switch to Newton. I verified the output match the same type of run with AMG.

I'll make a few comments after creating the pull request to clarify some rearranging I had to do for the StokesMatrixFree class.

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

tcclevenger commented on include/aspect/stokes_matrix_free.h:
```diff
@@ -306,8 +306,21 @@ namespace aspect
       virtual void setup_dofs()=0;
 
       /**
```
I had to move `evaluate_material_model()` and `correct_stokes_rhs()` to be public functions since I now need to call them separate from the `build_preconditioner()` in `solver_schemes.cc` and `assembly.cc`. Also, now `build_preconditioner()` is the same as the old `compute_A_block_diagonals()`.

tjhei replied:
I like the new naming.

tcclevenger commented on source/simulator/assembly.cc:
```diff
@@ -814,6 +817,13 @@ namespace aspect
       }
 
 
```
It seems natural to me to update to matrix-free right-hand side right after assembling. Maybe the `evaluate_material_model()` can move, however. Where are the material model values set and reset for each nonlinear iteration? In `core.cc`? Maybe we don't need to do this every time we assemble.

tcclevenger commented on source/simulator/stokes_matrix_free.cc:
```diff
@@ -1322,7 +1322,6 @@ namespace aspect
                                sim.pressure_scaling,
                                sim.triangulation,
                                dof_handler_projection);
```
the mass matrix diagonal is needed for the preconditioner. It seems more appropriate in the `build_preconditioner()` function.

gassmoeller commented on source/simulator/stokes_matrix_free.cc:
```diff
@@ -2021,21 +2020,15 @@ namespace aspect
   }
 
 
-
```
Keep the third empty line, that is our usual style

gassmoeller commented on doc/modules/changes/20200214_clevenger:
```diff
@@ -0,0 +1,3 @@
+New: Geometric Multigrid now supports computations with Picard iteration.
```
```suggestion
New: The geometric multigrid preconditioner for the Stokes system now supports Picard iterations for the defect correction form of the Stokes equations.
```

The original message was somewhat misleading, because as far as I can tell Picard iterations for the original form of the Stokes equations were already supported beforehand.

tcclevenger replied:
Yes, this is correct!

gassmoeller commented:
/rebuild

tcclevenger commented:
> Looks good, just a few minor comments.

Thanks, @gassmoeller! I've addressed the comments. I'll make sure the tests pass (last time I think one failed? We will see.)

gassmoeller commented:
Thanks. 

> last time I think one failed?

You can also check old test results by clicking on the little red x next to the commit hash in the pull request conversation (right above your comment). Saves you some waiting for the tester.

tcclevenger commented:
As with all the other GMG tests, the test fails because of the dependence on the machine architecture. I tested and verified on an AVX machine (vectorization level 2) and the Jenkins tester is a SSE2 machine (vectorization level 1). This can slightly alter iteration counts (as I see from the test output). Is there a way for me to access the output from Jenkins to change my test output file?

tjhei commented:
Details, artifacts in the top right.

tcclevenger commented:
Issues with the test have been resolved. Everything passed. Let me know if you have anymore comments!
